Source/PluginProcessor.h

```diff
@@ -72,17 +74,39 @@ class SoomplerAudioProcessor  : public AudioProcessor, ChangeListener
 
     void setTransportStateListener(TransportStateListener*);
 
-    double getCurrentAudioPosition() const;
+    double getCurrentAudioPosition();
+
+    int64 getTotalLengthOfSample() const {
+        return transportSource.getTotalLength();
+    }
+
+    void updateTransportState();
 
     void processTransport(AudioBuffer<float>& buffer);
 
+    // percent of a sample length
+    void setSampleStartPosition(int64 sample);
+    void setSampleEndPosition(int64 sample);
+
+    void setVolume(double volume) {
+        this->volume = volume;
+        transportSource.setGain(volume);
+    }
+
+    void notifyTransportStateChanged(TransportState state);
+
+    std::vector<int> getActiveNotes();

```

Comment:
vector will be copied each time. reference

Reply:
ok